### ISSUE TITLE
Alteon image Upload

### ISSUE THREAD

**Issue by maxaschenbrenner** [Radware/radware-ansible#49]
##### ENVIRONMENT
ansible-playbook 2.10.4
python version = 3.6.8
Software version: 32.4.7.50 (Alteon)
Distribution: VA

##### SEVERITY
High

##### SUMMARY
used example from https://github.com/Radware/radware-ansible/blob/master/plugins/modules/alteon_software_image_upload.py
Execution fails at "generate_pass":
1) when "generate_pass: false" 
Alteon returns "status 406" 
via verbose:  
"status": "err", "message": "406 Not Acceptable", "application message": "9/7/21 13:28 Image downloaded via Browser  - Invalid Upgrade Password"
-> when manually uploading a software via SSH, it will work and there's no need for an upgrade-password (with alteon Software v32)
2) when "generate_pass: true" 
Alteon tries to establish connection to receive upgrade-password. As the VA is not internet-connected, this fails
via verbose:
"Name or service not known url= http://services.radware.com:80/api/securityupdate/v1/password"

Also please note that there is a typo @ _https://github.com/Radware/radware-ansible/blob/master/plugins/modules/alteon_software_image_upload.py_ in the Example: "radware.radware_modules.alteo_software_image_upload:"

##### STEPS TO REPRODUCE
- hosts: ALTEON-NAME
  connection: local
  vars:
    ansible_python_interpreter: "{{ ansible_playbook_python }}"
    vdirect_validate_certs: no
  vars_prompt:
    - name: ansible_password
      prompt: "Password"
      private: yes
  tasks:
    - name: alteon device software upload
      radware.radware_modules.alteon_software_image_upload:
        provider:
          server: "{{ansible_host}}"
          user: "{{ansible_user}}"
          password: "{{ansible_password}}"
          validate_certs: no
          ssh_port: 22
          timeout: 5
           adc_slot: 2
        generate_pass: true
        timeout_seconds: 700
        file_path: IMAGE-FILE-PATH+NAME

##### EXPECTED RESULTS
successful upload of specified image to target device(s)

##### ACTUAL RESULTS
**generate_pass=true**
PLAY [ALTEON-NAME] ************************************************************************************************************************
TASK [alteon device software upload] ******************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Details: None.
[WARNING]: Module did not set no_log for generate_pass
[WARNING]: Module did not set no_log for password
fatal: [ALTEON-NAME]: FAILED! => {"changed": false, "msg": "\nRest Request Error:\nStatus: None\nReason: <urllib3.connection.HTTPConnection object at 0x7f6df6421d30>: Failed to establish a new connection: [Errno -2] Name or service not known\nurl= http://services.radware.com:80/api/securityupdate/v1/password\nmethod= POST\nbody= b'{\"mac\": \"6cb0f5d34d3c\", \"filesize\": 1169413151}'\nDetails: None."}
PLAY RECAP ********************************************************************************************************************************************
ALTEON-NAME    : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0  

**generate_pass=false**
PLAY [ALTEON-NAME] ************************************************************************************************************************
TASK [alteon device software upload] ******************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Alteon
[WARNING]: Module did not set no_log for generate_pass
[WARNING]: Module did not set no_log for password
fatal: [ALTEON-NAME]: FAILED! => {"changed": false, "msg": "\nRest Request Error:\nStatus: 406\nReason: None\nurl= https://ALTEON-IP-redacted:443/config/softwareimport?type=adc&adcimg=2\nmethod= POST\nbody= None\nDetails: \r\n{\n\t\"status\": \"err\",\n\t\"message\": \"406 Not Acceptable\",\n\t\"application message\": \"9/7/21 14:3 Image downloaded via Browser  - Invalid Upgrade Password\"\n}\r\n.\nAlteon"}
PLAY RECAP ********************************************************************************************************************************************
ALTEON-NAME    : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

**Comment by rdwrtac** [comment]
Hello  maxaschenbrenner,

I understand the first scenario mentioned when "generate_pass: false", and I would verify it. Before that, please clarify the second scenario when "generate_pass: true" as it is not clear what is exactly expected.

Because for "generate_pass: true" when Alteon doesn't have access to the internet, the Alteon won't be able to generate a password. This is the general operation for password upgrade and not Ansible specific. I will appreciate clarification on it.

In addition, could you please tell me where the typo? As I could not identify it.

Best regards,
rdwrtac

**Comment by maxaschenbrenner** [comment]
Dear rdwrtac,

the second scenario ("generate_pass: true") is okay - and totally expected. I just wanted to mention it, to make clear, that this is not a viable option for us. So with the other option not working, that doesn't leave me much...
Regarding the Typo: 
"radware.radware_modules.**alteo**_software_image_upload:"
it's missing an "n" - it needs to be "alteon" to work ;-)

**Comment by rdwrtac** [comment]
Hello  maxaschenbrenner,

Got it, now I understand. :-)
I'm checking this, and I will update you as soon as possible.

Best regards,
rdwrtac

**Comment by rdwrtac** [comment]
Hello maxaschenbrenner,

From my testing in the lab, the upgrade password field is mandatory when uploaded an image for a major version. When uploading an image for a minor version, the password is not required and can be left empty, this is the general behavior of Alteon.

I tested Major and Minor image uploads.

1. For a minor example, 32.4.5.50 to 32.4.7.50, I don't require any upgrade password, and after executing the playbook, the uploaded image can be found in Alteon.

++++++++++++++++++++++
The yaml file
++++++++++++++++++++++
- hosts: localhost

  tasks:
  - name: alteon device software
    radware.radware_modules.alteon_software_image_upload:
      provider: 
        server: 10.176.25.10
        user: admin
        password: admin
        validate_certs: no
        https_port: 443
        ssh_port: 22
        timeout: 5
      adc_slot: 2
      generate_pass: false
      timeout_seconds: 700
      file_path: /home/radware/Downloads/AlteonOS-32-4-7-50_rls_39.img

++++++++++++++++++++++
Result
++++++++++++++++++++++


[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [alteon device software] *********************************************************************************************************************************************************************************************************
[WARNING]: Module did not set no_log for generate_pass
[WARNING]: Module did not set no_log for password
changed: [localhost]

PLAY RECAP ****************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


2. For a major version example, 32.4.5.50 to 32.6. 5.50 I perform two different scenarios—one with an upgrade password and one without an upgrade password.

The "password: <Upgrade_Password_String>" field was removed when testing without password.

++++++++++++++++++++++
The yaml file
++++++++++++++++++++++

- hosts: localhost

  tasks:
  - name: alteon device software
    radware.radware_modules.alteon_software_image_upload:
      provider: 
        server: 10.176.25.10
        user: admin
        password: admin
        validate_certs: no
        https_port: 443
        ssh_port: 22
        timeout: 5
      adc_slot: 2
      generate_pass: false
      password: <Upgrade_Password_String>
      timeout_seconds: 700
      file_path: /home/radware/Downloads/AlteonOS-32-6-5-50_rls_49.img

++++++++++++++++++++++
 Fails to upload the image when upgrade password is missing
++++++++++++++++++++++


PLAY [localhost] **********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [alteon device software] *********************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Alteon
[WARNING]: Module did not set no_log for generate_pass
[WARNING]: Module did not set no_log for password
fatal: [localhost]: FAILED! => {"changed": false, "msg": "\nRest Request Error:\nStatus: 406\nReason: None\nurl= https://10.176.25.10:443/config/softwareimport?type=adc&adcimg=2\nmethod= POST\nbody= None\nDetails: \r\n{\n\t\"status\": \"err\",\n\t\"message\": \"406 Not Acceptable\",\n\t\"application message\": \"9/13/21 5:15 Image downloaded via Browser  - Invalid Upgrade Password\"\n}\r\n.\nAlteon"}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   



++++++++++++++++++++++
Successfully uploading the image when the upgrade password field is added.
++++++++++++++++++++++

[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [alteon device software] *********************************************************************************************************************************************************************************************************
[WARNING]: Module did not set no_log for generate_pass
[WARNING]: Module did not set no_log for password
changed: [localhost]

PLAY RECAP ****************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   



In conclusion, based on all the testing I performed, I didn't find a problem with the module's operation, and the only minor problem involved in this module is the typo in the name.


Best regards,
rdwrtac

**Comment by yaronantebi** [comment]
Please find below example for image upgrade - 

- hosts: localhost
  vars_files:
    - vars/alteon_params.yml
  tasks:
    - name: alteon device software
      radware.radware_modules.alteon_software_image_upload:
        provider: "{{ radware_provider }}"
        adc_slot: 2
        generate_pass: false
        timeout_seconds: 700
        file_path: /home/radware/images/AlteonOS-32-4-7-0_rls_112.img
